File: core/src/main/kotlin/com/pokerface/core/Equity.kt
package com.pokerface.core

import kotlin.random.Random

data class EquityResult(
    val winProbability: Double,
    val tieProbability: Double,
    val loseProbability: Double,
    val trials: Int,
) {
    /** 무승부를 절반의 승리로 환산한 실질 승률 (팟 오즈 비교에 사용). */
    val equity: Double get() = winProbability + tieProbability / 2.0
}

/**
 * 둘이서(헤즈업) 플레이 시, 상대 홀카드는 알 수 없다고 가정하고
 * 남은 덱에서 균등 무작위로 뽑는다는 전제로 승률을 추정한다.
 *
 * 리버(보드 5장 완성)에서는 상대 홀카드 조합을 전수 계산해 정확한 승률을 낸다.
 * 그 이전 스트리트에서는 몬테카를로 시뮬레이션으로 추정한다.
 */
object EquityCalculator {

    fun estimate(
        heroHole: List<Card>,
        board: List<Card>,
        trials: Int = 15_000,
        random: Random = Random(System.nanoTime()),
    ): EquityResult {
        require(heroHole.size == 2) { "홀 카드는 2장이어야 합니다" }
        require(board.size in 0..5) { "보드는 0~5장이어야 합니다" }

        val known = heroHole + board
        val deck = Deck.remaining(known)
        val boardNeeded = 5 - board.size

        return if (boardNeeded == 0) {
            exactRiverEquity(heroHole, board, deck)
        } else {
            monteCarloEquity(heroHole, board, deck, boardNeeded, trials, random)
        }
    }

    private fun exactRiverEquity(heroHole: List<Card>, board: List<Card>, deck: List<Card>): EquityResult {
        val heroValue = HandEvaluator.bestHand(heroHole + board)
        var win = 0
        var tie = 0
        var lose = 0
        for (i in deck.indices) {
            for (j in i + 1 until deck.size) {
                val villainValue = HandEvaluator.bestHand(listOf(deck[i], deck[j]) + board)
                when {
                    heroValue > villainValue -> win++
                    heroValue.compareTo(villainValue) == 0 -> tie++
                    else -> lose++
                }
            }
        }
        val total = win + tie + lose
        return EquityResult(win.toDouble() / total, tie.toDouble() / total, lose.toDouble() / total, total)
    }

    private fun monteCarloEquity(
        heroHole: List<Card>,
        board: List<Card>,
        deck: List<Card>,
        boardNeeded: Int,
        trials: Int,
        random: Random,
    ): EquityResult {
        val pool = deck.toMutableList()
        val drawCount = 2 + boardNeeded
        var win = 0
        var tie = 0
        var lose = 0

        repeat(trials) {
            partialShuffle(pool, drawCount, random)
            val villainHole = listOf(pool[0], pool[1])
            val fullBoard = if (boardNeeded > 0) board + pool.subList(2, 2 + boardNeeded) else board

            val heroValue = HandEvaluator.bestHand(heroHole + fullBoard)
            val villainValue = HandEvaluator.bestHand(villainHole + fullBoard)
            val cmp = heroValue.compareTo(villainValue)
            when {
                cmp > 0 -> win++
                cmp == 0 -> tie++
                else -> lose++
            }
        }

        return EquityResult(win.toDouble() / trials, tie.toDouble() / trials, lose.toDouble() / trials, trials)
    }

    /** [list]의 앞 [count]개 위치를 무작위 표본으로 채우는 부분 피셔-예이츠 셔플. */
    private fun <T> partialShuffle(list: MutableList<T>, count: Int, random: Random) {
        for (i in 0 until count) {
            val j = i + random.nextInt(list.size - i)
            val tmp = list[i]
            list[i] = list[j]
            list[j] = tmp
        }
    }
}

File: web/src/poker/equity.ts
import { Card } from "./card";
import { remainingDeck } from "./deck";
import { bestHand, compareHandValues } from "./handEvaluator";

export interface EquityResult {
  winProbability: number;
  tieProbability: number;
  loseProbability: number;
  trials: number;
}

/** 무승부를 절반의 승리로 환산한 실질 승률 (팟 오즈 비교에 사용). */
export function equityOf(result: EquityResult): number {
  return result.winProbability + result.tieProbability / 2;
}

/**
 * 둘이서(헤즈업) 플레이 시, 상대 홀카드는 알 수 없다고 가정하고
 * 남은 덱에서 균등 무작위로 뽑는다는 전제로 승률을 추정한다.
 *
 * 리버(보드 5장 완성)에서는 상대 홀카드 조합을 전수 계산해 정확한 승률을 낸다.
 * 그 이전 스트리트에서는 몬테카를로 시뮬레이션으로 추정한다.
 */
export function estimateEquity(
  heroHole: Card[],
  board: Card[],
  trials = 12000,
  rng: () => number = Math.random,
): EquityResult {
  if (heroHole.length !== 2) throw new Error("홀 카드는 2장이어야 합니다");
  if (board.length < 0 || board.length > 5) throw new Error("보드는 0~5장이어야 합니다");

  const deck = remainingDeck([...heroHole, ...board]);
  const boardNeeded = 5 - board.length;

  if (boardNeeded === 0) {
    return exactRiverEquity(heroHole, board, deck);
  }
  return monteCarloEquity(heroHole, board, deck, boardNeeded, trials, rng);
}

function exactRiverEquity(heroHole: Card[], board: Card[], deck: Card[]): EquityResult {
  const heroValue = bestHand([...heroHole, ...board]);
  let win = 0, tie = 0, lose = 0;
  for (let i = 0; i < deck.length; i++) {
    for (let j = i + 1; j < deck.length; j++) {
      const villainValue = bestHand([deck[i], deck[j], ...board]);
      const cmp = compareHandValues(heroValue, villainValue);
      if (cmp > 0) win++;
      else if (cmp === 0) tie++;
      else lose++;
    }
  }
  const total = win + tie + lose;
  return { winProbability: win / total, tieProbability: tie / total, loseProbability: lose / total, trials: total };
}

function monteCarloEquity(
  heroHole: Card[],
  board: Card[],
  deck: Card[],
  boardNeeded: number,
  trials: number,
  rng: () => number,
): EquityResult {
  const pool = [...deck];
  const drawCount = 2 + boardNeeded;
  let win = 0, tie = 0, lose = 0;

  for (let t = 0; t < trials; t++) {
    partialShuffle(pool, drawCount, rng);
    const villainHole = [pool[0], pool[1]];
    const fullBoard = boardNeeded > 0 ? [...board, ...pool.slice(2, 2 + boardNeeded)] : board;

    const heroValue = bestHand([...heroHole, ...fullBoard]);
    const villainValue = bestHand([...villainHole, ...fullBoard]);
    const cmp = compareHandValues(heroValue, villainValue);
    if (cmp > 0) win++;
    else if (cmp === 0) tie++;
    else lose++;
  }

  return { winProbability: win / trials, tieProbability: tie / trials, loseProbability: lose / trials, trials };
}

/** [list]의 앞 [count]개 위치를 무작위 표본으로 채우는 부분 피셔-예이츠 셔플. */
function partialShuffle<T>(list: T[], count: number, rng: () => number): void {
  for (let i = 0; i < count; i++) {
    const j = i + Math.floor(rng() * (list.length - i));
    const tmp = list[i];
    list[i] = list[j];
    list[j] = tmp;
  }
}
